File: report/src/main/java/com/vmlens/report/runelementtype/objecthashcodemap/ObjectHashCodeMap.java
package com.vmlens.report.runelementtype.objecthashcodemap;

import java.util.HashMap;

public class ObjectHashCodeMap {

    private int maxCode;

    private final HashMap<Long,ThreadIndices> hashCodeToThreadIndices = new HashMap<>();

    public void add(long objectHashCode, int threadIndex) {
        ThreadIndices indices = hashCodeToThreadIndices.get(objectHashCode);
        if(indices == null) {
            hashCodeToThreadIndices.put(objectHashCode,new OneThreadIndex(threadIndex));
        } else  {
            hashCodeToThreadIndices.put(objectHashCode,indices.addThreadIndex(threadIndex,this));
        }
    }

    public String asUiString(long objectHashCode) {
        return hashCodeToThreadIndices.get(objectHashCode).asUiString();
    }

    public int newCode() {
         maxCode++;
        return maxCode;
    }
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/objecthashcodemap/ObjectHashCodeMap.scala
package com.anarsoft.race.detection.report.element.runelementtype.objecthashcodemap

import java.util.HashMap


class ObjectHashCodeMap {
  private var maxCode = 0
  final private val hashCodeToThreadIndices = new HashMap[Long, ThreadIndices]

  def add(objectHashCode: Long, threadIndex: Int): Unit = {
    val indices = hashCodeToThreadIndices.get(objectHashCode)
    if (indices == null) hashCodeToThreadIndices.put(objectHashCode, new OneThreadIndex(threadIndex))
    else hashCodeToThreadIndices.put(objectHashCode, indices.addThreadIndex(threadIndex, this))
  }

  def asUiString(objectHashCode: Long): String = hashCodeToThreadIndices.get(objectHashCode).asUiString
  
  
  def isSingleThreaded(objectHashCode: Long) : Boolean =  hashCodeToThreadIndices.get(objectHashCode).isSingleThreaded
  
  def id(objectHashCode: Long): Option[Int] = {
    hashCodeToThreadIndices.get(objectHashCode).id;
  }

  def newCode: Int = {
    maxCode += 1
    maxCode
  }
}
